test/integration/deep/opaqueports/opaque_ports_test.go

```diff
@@ -14,171 +14,374 @@ import (
+				),
+				appName:   opaquePodApp,
+				appChecks: checks(hasInboundTCPTraffic),
+			},
+			{
+				name:   "calling a meshed service when opaque annotation is on calling pod",
+				scName: opaqueSvcSC,
+				scChecks: checks(
+					hashNoOutbondHTTPRequest,
+					hasOutboundTCPWithTLSAndAuthority,
+				),
+				appName:   opaqueSvcApp,
+				appChecks: checks(hasInboundTCPTraffic),
+			},
+			{
+				name:   "calling an unmeshed service",
+				scName: opaqueUnmeshedSvcSC,
+				scChecks: checks(
+					hashNoOutbondHTTPRequest,
+					hasOutboundTCPWithAuthorityAndNoTLS,
+				),
+			},
+		}
+
+		for _, tc := range tests {
+			t.Run(tc.name, func(t *testing.T) {
+				err := TestHelper.RetryFor(30*time.Second, func() error {
+					if err := checkPodMetrics(ctx, opaquePortsNs, tc.scName, tc.scChecks); err != nil {
+						return fmt.Errorf("failed to check metrics for client pod:%w", err)
+					}
+					if tc.appName == "" {
+						return nil
+					}
+					if err := checkPodMetrics(ctx, opaquePortsNs, tc.appName, tc.appChecks); err != nil {
+						return fmt.Errorf("failed to check metrics for app pod:%w", err)
+					}
+					return nil
+				})
 				if err != nil {
-					return fmt.Errorf("error getting metrics for pod\n%w", err)
+					testutil.AnnotatedFatalf(t, "unexpected metric for client pod", "unexpected metric for client pod: %s", err)
 				}
-				if !tcpMetricRE.MatchString(metrics) {
-					return fmt.Errorf("failed to find expected TCP metric when pod is opaque\n%s", metrics)
-				}
-
-				return nil
 			})
+		}
+	})
+}
 
-			if err != nil {
-				testutil.AnnotatedFatalf(t, "unexpected metric output", "unexpected metric output: %s", err)
-			}
-		})
-
-		t.Run("expect inbound TCP connection metric with expected TLS identity for opaque service app", func(t *testing.T) {
-			// Check the slow cooker metrics
-			err := TestHelper.RetryFor(30*time.Second, func() error {
-				pods, err := TestHelper.GetPods(ctx, opaquePortsNs, map[string]string{"app": opaqueSvcSC})
-				if err != nil || len(pods) == 0 {
-					return fmt.Errorf("error getting pods\n%w", err)
-				}
-				metrics, err := getPodMetrics(pods[0], opaquePortsNs)
-				if err != nil {
-					return fmt.Errorf("error getting metrics for pod\n%w", err)
-				}
-				if httpRequestTotalMetricRE.MatchString(metrics) {
-					return fmt.Errorf("expected not to find HTTP outbound requests when service is opaque\n%s", metrics)
-				}
-				// Check the application metrics
-				pods, err = TestHelper.GetPods(ctx, opaquePortsNs, map[string]string{"app": opaqueSvcApp})
-				if err != nil {
-					return fmt.Errorf("error getting pods\n%w", err)
-				}
-				metrics, err = getPodMetrics(pods[0], opaquePortsNs)
-				if err != nil {
-					return fmt.Errorf("error getting metrics for pod\n%w", err)
-				}
-				if !tcpMetricRE.MatchString(metrics) {
-					return fmt.Errorf("failed to find expected TCP metric when pod is opaque\n%s", metrics)
-				}
+func TestOpaquePortsCalledByServiceIP(t *testing.T) {
```

Comment:
The difference between these two is that one uses service name and the other uses service IP. So if we have:
```
> kubectl get svc -n linkerd-opaque-ports-called-by-service-name-test
NAME                          TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)                  AGE
svc-opaque-service   ClusterIP   10.43.223.137   <none>        8080/TCP                 11
```
in one test SC will call `svc-opaque-service:8080` and the other will call `10.43.223.137:8080`. 

If you say that `CalledByServiceIP` is redundant, then I am willing to remove it. However, my personal comment is that it felt natural for me to have this test, as another way a pod can call a service. And when you say:

>Whether the target is named or unnamed (e.g IP) shouldn't make a difference 

my first instinct is that this is what tests are for: to make sure we are right about our assumptions.

Having said all that, I don't know that much about linkerd and name resolution. I was a little surprised that DNS is involved in this scenario. 🙃 So if you feel like this test is redundant and this stuff is tested somewhere else, then I am all willing to remove it.

Reply:
ok, I was thinking about it. What you are saying is that resolution of `svc-opaque-service` to `10.43.223.137` will happen before app will "talk to" linkerd sidecar, right?  If I get this right, I think we should get rid of this test.

Reply:
@krzysztofdrys 

> What you are saying is that resolution of svc-opaque-service to 10.43.223.137 will happen before app will "talk to" linkerd sidecar, right?

Yup. This wouldn't make a difference in how opaque ports are handled. We can just use the fqn since it's easier to read in tests (and we don't need any logic to get the service's IP).